src/circuit.cpp

```diff
@@ -1,13 +1,331 @@
+// It basically boils down to vector vs. array - SSZ implementations work with vectors with max size bounded by VALIDATORS_MAX_SIZE,
+// but only carrying actual number of validators (~1M by Oct 2023), while circuit has to use array and thus would carry a full
+// VALIDATORS_MAX_SIZE elements - which is a lot. So we're using a smaller "problem size" to carry (and pass aroung) the necessary
+// minimum of "padding" elemetns.
+constexpr std::size_t PROBLEM_SIZE_LOG2 = 4;
+constexpr std::size_t VALIDATORS_COUNT = 1 << PROBLEM_SIZE_LOG2; // 2 ** PROBLEM_SIZE_LOG2
+constexpr std::size_t VALIDATORS_TREE_HEIGHT = PROBLEM_SIZE_LOG2; // 2 ** PROBLEM_SIZE_LOG2
+constexpr std::size_t BALANCES_COUNT = 1 << PROBLEM_SIZE_LOG2; // 2 ** PROBLEM_SIZE_LOG2
+constexpr std::size_t BALANCES_TREE_HEIGHT = PROBLEM_SIZE_LOG2 - BALANCES_PER_LEAF_LOG2;
+constexpr std::size_t BALANCES_LEAFS_COUNT = BALANCES_COUNT / BALANCES_PER_LEAF;
+
+constexpr std::size_t VALIDATOR_FIELDS = 8;
+constexpr std::size_t BEACON_STATE_FIELD_INCLUSION_PROOF_LENGTH = 5;
+constexpr std::size_t VALIDATORS_FIELD_INDEX = 11;
+constexpr std::size_t BALANCES_FIELD_INDEX = 12;
+
+constexpr std::size_t BEACON_BLOCK_FIELDS_COUNT = 5;
+
+constexpr bool BYTE_ORDER_MSB = true;
+constexpr bool BYTE_ORDER_LSB = false;
+
+bool is_same(block_type block0, block_type block1){
+    return block0[0] == block1[0] && block0[1] == block1[1];
+}
+
+template <std::size_t LayerSize>
+block_type hash_layer(std::array<block_type, LayerSize> input, size_t layer){
+    constexpr size_t NextLayerSize = (LayerSize % 2 == 0) ? LayerSize / 2 : (LayerSize / 2 + 1);
+    std::array<block_type, NextLayerSize> next_layer;
+
+    for (std::size_t leaf_index = 0; leaf_index < LayerSize / 2; leaf_index++) {
+        next_layer[leaf_index] = hash<hash_type>(input[2 * leaf_index], input[2 * leaf_index + 1]);
+    }
+    if (LayerSize % 2 != 0) {
+        next_layer[NextLayerSize - 1] = hash<hash_type>(LayerSize - 1, precomputed_zero_hashes[layer]);
+    }
+    if (LayerSize == 2)
+        return next_layer[0];
+    else
+        return hash_layer<NextLayerSize>(next_layer, layer + 1);
+}
+
+template <std::size_t LayerSize>
+block_type hash_tree(std::array<block_type, LayerSize> input){
+    return hash_layer(input, 0);
+}
+
+template <std::size_t BalancesCount>
+std::array<block_type, BalancesCount / BALANCES_PER_LEAF> pack_balances_into_field_elements(
+    std::array<uint64_t, BalancesCount> validator_balances
+) {
+    std::array<block_type, BALANCES_LEAFS_COUNT> leafs;
+    for (std::size_t i = 0; i < BALANCES_LEAFS_COUNT; i++) {
+        // MSB first
+        decomposed_int64_type first_balance_in_block_bits =
+            __builtin_assigner_bit_decomposition64(validator_balances[4*i], BYTE_ORDER_LSB);
+        decomposed_int64_type second_balance_in_block_bits =
+            __builtin_assigner_bit_decomposition64(validator_balances[4*i+1], BYTE_ORDER_LSB);
```

Comment:
Just double checking, if last argument is true, it means "most-significant-bytes-first" aka "big-endian"?

So if we need little-endian, third argument should be false?

Reply:
True is MSB, right.